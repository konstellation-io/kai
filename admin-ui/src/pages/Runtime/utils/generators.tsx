import React from 'react';
import Button from '../../../components/Button/Button';
import StartIcon from '@material-ui/icons/SkipNext';
import StopIcon from '@material-ui/icons/PauseCircleFilled';
import PublishIcon from '@material-ui/icons/PlayCircleFilledOutlined';
import UnpublishIcon from '@material-ui/icons/Block';
import { VersionStatus } from '../../../graphql/types/globalTypes';

function generateActionButton(label: string, icon: any, action: Function) {
  return (
    <Button
      key={label}
      label={label}
      onClick={action}
      height={30}
      Icon={icon}
    />
  );
}

function getStateToButtons(
  publishAction: Function,
  startAction: Function,
  stopAction: Function,
  unpublishAction: Function
) {
  const buttonStart = generateActionButton('START', StartIcon, startAction);
  const buttonStop = generateActionButton('STOP', StopIcon, stopAction);
  const buttonPublish = generateActionButton(
    'PUBLISH',
    PublishIcon,
    publishAction
  );
  const buttonUnpublish = generateActionButton(
    'UNPUBLISH',
    UnpublishIcon,
    unpublishAction
  );

  return {
    [VersionStatus.STOPPED]: [buttonStart],
    [VersionStatus.PUBLISHED]: [buttonUnpublish],
    [VersionStatus.STARTED]: [buttonPublish, buttonStop],
    [VersionStatus.STOPPED]: [buttonStart]
  };
}

export function getVersionActionButtons(
  publishAction: Function,
  startAction: Function,
  stopAction: Function,
  unpublishAction: Function,
  status?: string
) {
  const stateToButtons: { [key: string]: any } = getStateToButtons(
    publishAction,
    startAction,
    stopAction,
    unpublishAction
  );

  return stateToButtons[status || ''];
}
