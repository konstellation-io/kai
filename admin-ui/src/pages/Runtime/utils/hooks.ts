import { useMutation } from '@apollo/react-hooks';
import {
  ACTIVATE_VERSION,
  DEACTIVATE_VERSION,
  DEPLOY_VERSION,
  STOP_VERSION,
  ActivateVersionResponse,
  DeactivateVersionResponse,
  DeployVersionResponse,
  StopVersionResponse,
  VersionActionVars
} from '../pages/RuntimeStatusPreview/RuntimeStatusPreview.graphql';

export enum versionActions {
  activate = 'activateVersion',
  deploy = 'deployVersion',
  stop = 'stopVersion',
  deactivate = 'deactivateVersion'
}

export default function useVersionAction() {
  const [activateMutation, { loading: loadingM1 }] = useMutation<
    ActivateVersionResponse,
    VersionActionVars
  >(ACTIVATE_VERSION);
  const [deactivateMutation, { loading: loadingM2 }] = useMutation<
    DeactivateVersionResponse,
    VersionActionVars
  >(DEACTIVATE_VERSION);
  const [deployMutation, { loading: loadingM3 }] = useMutation<
    DeployVersionResponse,
    VersionActionVars
  >(DEPLOY_VERSION);
  const [stopMutation, { loading: loadingM4 }] = useMutation<
    StopVersionResponse,
    VersionActionVars
  >(STOP_VERSION);

  const mutationLoading = [loadingM1, loadingM2, loadingM3, loadingM4].some(
    el => el
  );

  function getMutationVars(versionId: string, comment?: string) {
    const variables = {
      variables: {
        input: {
          versionId: versionId
        }
      }
    } as any;

    if (comment) {
      variables.variables.input.comment = comment;
    }

    return variables;
  }

  return {
    [versionActions.activate]: activateMutation,
    [versionActions.deploy]: deployMutation,
    [versionActions.stop]: stopMutation,
    [versionActions.deactivate]: deactivateMutation,
    getMutationVars,
    mutationLoading
  };
}
