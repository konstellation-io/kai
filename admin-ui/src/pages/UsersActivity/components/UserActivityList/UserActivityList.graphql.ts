import gql from 'graphql-tag';
import { UserActivity } from '../../../../graphql/models';

export interface UserActivityResponse {
  userActivityList: UserActivity[];
}

export const GET_USERS_ACTIVITY = gql`
  query GetUsersActivity {
    userActivityList {
      user {
        email
      }
      date
      type
      vars {
        key
        value
      }
    }
  }
`;
