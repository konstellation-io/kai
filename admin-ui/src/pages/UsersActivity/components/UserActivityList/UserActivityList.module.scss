@import '../../../../styles/mixins';
@import '../../../../styles/colors';

.row {
  @include font-body;

  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding: 10px 35px;
  border-bottom: solid 2px #15161a;
}

.userWithIcon,
.date {
  display: flex;
  flex-direction: row;
  line-height: 14px;

  svg {
    line-height: 20px;
  }
}
.user {
  color: white;
  margin: 0 0 0 10px;
}

.message {
  margin: 5px 0 0 25px;
  padding-right: 45px;

  a {
    color: $font-color;
    transition: color ease 0.3s;

    &:hover {
      color: $font-color-white;
    }
  }
}

.comment {
  color: $font-color-dark;

  margin: 5px 0 0 25px;
  padding-right: 45px;
}

.date {
  white-space: nowrap;

  p {
    margin: 0 0 0 10px;
  }
}
