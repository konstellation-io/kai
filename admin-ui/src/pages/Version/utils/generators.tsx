import React, { ReactElement, MouseEvent, FunctionComponent } from 'react';
import Button from '../../../components/Button/Button';
import StartIcon from '@material-ui/icons/SkipNext';
import StopIcon from '@material-ui/icons/PauseCircleFilled';
import PublishIcon from '@material-ui/icons/PlayCircleFilledOutlined';
import UnpublishIcon from '@material-ui/icons/Block';
import { VersionStatus } from '../../../graphql/types/globalTypes';
import { SvgIconProps } from '@material-ui/core/SvgIcon';

function generateActionButton(
  label: string,
  icon: FunctionComponent<SvgIconProps>,
  action: (e: MouseEvent<HTMLDivElement>) => void
): ReactElement {
  return (
    <Button
      key={label}
      label={label}
      onClick={action}
      height={30}
      Icon={icon}
    />
  );
}

function getStateToButtons(
  publishAction: (e: MouseEvent<HTMLDivElement>) => void,
  startAction: (e: MouseEvent<HTMLDivElement>) => void,
  stopAction: (e: MouseEvent<HTMLDivElement>) => void,
  unpublishAction: (e: MouseEvent<HTMLDivElement>) => void
): { [key: string]: ReactElement[] } {
  const buttonStart = generateActionButton('START', StartIcon, startAction);
  const buttonStop = generateActionButton('STOP', StopIcon, stopAction);
  const buttonPublish = generateActionButton(
    'PUBLISH',
    PublishIcon,
    publishAction
  );
  const buttonUnpublish = generateActionButton(
    'UNPUBLISH',
    UnpublishIcon,
    unpublishAction
  );

  return {
    [VersionStatus.STOPPED]: [buttonStart],
    [VersionStatus.PUBLISHED]: [buttonUnpublish],
    [VersionStatus.STARTED]: [buttonPublish, buttonStop],
    [VersionStatus.STOPPED]: [buttonStart]
  };
}

export function getVersionActionButtons(
  publishAction: (e: MouseEvent<HTMLDivElement>) => void,
  startAction: (e: MouseEvent<HTMLDivElement>) => void,
  stopAction: (e: MouseEvent<HTMLDivElement>) => void,
  unpublishAction: (e: MouseEvent<HTMLDivElement>) => void,
  status?: string
) {
  const stateToButtons = getStateToButtons(
    publishAction,
    startAction,
    stopAction,
    unpublishAction
  );

  if (status === undefined) {
    return [];
  }

  return stateToButtons[status];
}
