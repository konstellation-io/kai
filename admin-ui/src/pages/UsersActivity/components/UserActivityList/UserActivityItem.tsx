import React from 'react';
import { formatDate } from '../../../../utils/format';
import getMessage from './messageGenerator';
import TimeIcon from '@material-ui/icons/AccessTime';
import ActivityIcon from '../ActivityIcon/ActivityIcon';
import styles from './UserActivityList.module.scss';
import { GetUsersActivity_userActivityList } from '../../../../graphql/queries/types/GetUsersActivity';

type Props = {
  userActivity: GetUsersActivity_userActivityList;
  idx: number;
};
function UserActivityItem({ userActivity, idx }: Props) {
  const [message, comment] = getMessage(userActivity);

  return (
    <div className={styles.row} key={`${userActivity.date}-${idx}`}>
      <div className={styles.date}>
        <div className={styles.dateIcon}>
          <TimeIcon className="icon-small" />
        </div>
        {formatDate(new Date(userActivity.date), true)}
      </div>
      <div className={styles.activityIcon}>
        <ActivityIcon activityType={userActivity.type} />
      </div>
      <div className={styles.info}>
        <p className={styles.user}>{userActivity.user.email}</p>
        <div className={styles.message}>{message}</div>
        {comment && <p className={styles.comment}>{comment}</p>}
      </div>
    </div>
  );
}

export default UserActivityItem;
