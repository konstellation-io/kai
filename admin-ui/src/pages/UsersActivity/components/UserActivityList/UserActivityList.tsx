import React from 'react';
import { formatDate } from '../../../../utils/format';

import EmailIcon from '@material-ui/icons/Email';
import TimeIcon from '@material-ui/icons/AccessTime';

import styles from './UserActivityList.module.scss';

import { UserActivity } from '../../../../graphql/models';

const typeToMessage: { [key: string]: string } = {
  LOGIN: 'Has logged in',
  LOGOUT: 'Has logged out',
  CREATE_RUNTIME: 'Has created a Runtime'
};

type Props = {
  data?: UserActivity[];
};

function UserActivityList({ data }: Props) {
  const usersActivity =
    data &&
    data.map((userActivity: UserActivity, idx: number) => {
      return (
        <div
          className={styles.row}
          key={`userActivityListElement${idx}`}
          data-testid={`userActivityListElement${idx}`}
        >
          <div className={styles.userAndMessage}>
            <span className={styles.userWithIcon}>
              <EmailIcon className="icon-regular" />
              <p className={styles.user}>{userActivity.user.email}</p>
            </span>
            <p className={styles.message}>{typeToMessage[userActivity.type]}</p>
          </div>
          <div className={styles.date}>
            <TimeIcon className="icon-regular" />
            <p>{formatDate(new Date(userActivity.date), true)}</p>
          </div>
        </div>
      );
    });

  return <>{usersActivity}</>;
}

export default UserActivityList;
