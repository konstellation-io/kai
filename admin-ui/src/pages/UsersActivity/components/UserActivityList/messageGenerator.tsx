import React from 'react';
import { Link } from 'react-router-dom';
import * as ROUTES from '../../../../constants/routes';
import { buildRoute } from '../../../../utils/routes';
import {
  UserActivity,
  UserActivityVar,
  UserActivityType
} from '../../../../graphql/models';

enum VarTypes {
  RUNTIME_ID = 'RUNTIME_ID',
  RUNTIME_NAME = 'RUNTIME_NAME',
  VERSION_ID = 'VERSION_ID',
  VERSION_NAME = 'VERSION_NAME',
  OLD_ACTIVE_VERSION_NAME = 'OLD_ACTIVE_VERSION_NAME',
  OLD_ACTIVE_VERSION_ID = 'OLD_ACTIVE_VERSION_ID',
  COMMENT = 'COMMENT',
  SETTING_NAME = 'SETTING_NAME',
  OLD_VALUE = 'OLD_VALUE',
  NEW_VALUE = 'NEW_VALUE',
  CONFIG_KEYS = 'CONFIG_KEYS'
}

function getVar(text: any) {
  return <strong>{text}</strong>;
}

export default function getMessage(
  userActivity: UserActivity
): [any, string | undefined] {
  let message: any = '';

  const vars = userActivity.vars.reduce(
    (result: { [key: string]: string }, v: UserActivityVar) => {
      result[v.key] = v.value;
      return result;
    },
    {}
  );

  let runtimeName: string | undefined = vars[VarTypes.RUNTIME_NAME],
    runtimeId: string | undefined = vars[VarTypes.RUNTIME_ID],
    versionName: string | undefined = vars[VarTypes.VERSION_NAME],
    versionId: string | undefined = vars[VarTypes.VERSION_ID],
    oldActiveVersionName: string | undefined =
      vars[VarTypes.OLD_ACTIVE_VERSION_NAME],
    oldActiveVersionId: string | undefined =
      vars[VarTypes.OLD_ACTIVE_VERSION_ID],
    comment: string | undefined = vars[VarTypes.COMMENT],
    settingName: string | undefined = vars[VarTypes.SETTING_NAME],
    oldValue: string | undefined = vars[VarTypes.OLD_VALUE],
    newValue: string | undefined = vars[VarTypes.NEW_VALUE],
    configKeys: string | undefined = vars[VarTypes.CONFIG_KEYS];

  const runtimeLink =
    userActivity.type === UserActivityType.CREATE_RUNTIME && runtimeId ? (
      <Link to={buildRoute.runtime(ROUTES.RUNTIME, runtimeId)}>
        {runtimeName}
      </Link>
    ) : (
      undefined
    );
  const versionLink =
    runtimeId && versionId ? (
      <Link
        to={buildRoute.version(
          ROUTES.RUNTIME_VERSION_STATUS,
          runtimeId,
          versionId
        )}
      >
        {`${runtimeName} - ${versionName}`}
      </Link>
    ) : (
      undefined
    );
  const oldVersionLink =
    userActivity.type === UserActivityType.ACTIVATE_VERSION &&
    runtimeId &&
    oldActiveVersionId ? (
      <Link
        to={buildRoute.version(
          ROUTES.RUNTIME_VERSION_STATUS,
          runtimeId,
          oldActiveVersionId
        )}
      >
        {oldActiveVersionName}
      </Link>
    ) : (
      undefined
    );

  switch (userActivity.type) {
    case UserActivityType.LOGIN:
      message = <>{'Has logged in'}</>;
      break;
    case UserActivityType.LOGOUT:
      message = <>{'Has logged out'}</>;
      break;
    case UserActivityType.CREATE_RUNTIME:
      message = (
        <>
          {'Has created a new Runtime: '}
          {getVar(runtimeLink)}
        </>
      );
      break;
    case UserActivityType.CREATE_VERSION:
      message = (
        <>
          {'Has created a new Version: '}
          {getVar(versionLink)}
        </>
      );
      break;
    case UserActivityType.ACTIVATE_VERSION:
      message = (
        <>
          {'Has activated version '}
          {getVar(versionLink)}
          {'. '}
          {oldVersionLink && (
            <>
              {'Previous active version: '}
              {getVar(oldVersionLink)}
            </>
          )}
        </>
      );
      break;
    case UserActivityType.DEACTIVATE_VERSION:
      message = (
        <>
          {'Has deactivated version '}
          {getVar(versionLink)}
        </>
      );
      break;
    case UserActivityType.STOP_VERSION:
      message = (
        <>
          {'Has stopped version '}
          {getVar(versionLink)}
        </>
      );
      break;
    case UserActivityType.DEPLOY_VERSION:
      message = (
        <>
          {'Has deployed version '}
          {getVar(versionLink)}
        </>
      );
      break;
    case UserActivityType.UPDATE_SETTING:
      message = (
        <>
          {'Has updated '}
          {getVar(settingName)}
          {' setting from '}
          {getVar(oldValue)}
          {' to '}
          {getVar(newValue)}
        </>
      );
      break;
    case UserActivityType.UPDATE_VERSION_CONFIGURATION:
      message = (
        <>
          {'Has changed settings: '}
          {getVar(configKeys)}
          {' from version '}
          {getVar(versionLink)}
        </>
      );
      break;
    default:
      break;
  }

  return [message, comment];
}
