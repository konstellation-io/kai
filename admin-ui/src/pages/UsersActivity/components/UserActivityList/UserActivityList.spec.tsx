import React from 'react';
import UserActivityList from './UserActivityList';
import { formatDate } from '../../../../utils/format';
import { render, cleanup } from '@testing-library/react';

import { MockedProvider } from '@apollo/react-testing';
import wait from 'waait';
import { act } from 'react-dom/test-utils';

import { usersActivityMock } from '../../../../mocks/settings';
import '@testing-library/jest-dom/extend-expect';

const mockData = usersActivityMock.result.data.usersActivity;

const Component = <UserActivityList data={mockData} />;

afterEach(cleanup);

it('Render UserActivityList without crashing', () => {
  const { container } = render(Component);

  expect(container).toMatchSnapshot();
});

it('Shows right texts', async () => {
  const { queryByTestId } = render(Component);

  const userActivityNode0 = queryByTestId('userActivityListElement0');
  const userActivityNode1 = queryByTestId('userActivityListElement1');
  expect(userActivityNode0).not.toBeNull();

  const dateFormatted = formatDate(new Date('Nov 27, 2019, 4:28 PM'), true);
  const user0 = userActivityNode0 && userActivityNode0.querySelector('.user');
  const message0 =
    userActivityNode0 && userActivityNode0.querySelector('.message');
  const date1 = userActivityNode1 && userActivityNode1.querySelector('.date p');
  expect(user0 && user0.textContent).toBe('user1@domain.com');
  expect(message0 && message0.textContent).toBe('Has logged in');
  expect(date1 && date1.textContent).toBe(dateFormatted);
});
