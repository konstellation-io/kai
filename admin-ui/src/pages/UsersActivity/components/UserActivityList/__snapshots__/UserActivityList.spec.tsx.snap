// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render UserActivityList without crashing 1`] = `
<div>
  <div
    class="row"
    data-testid="userActivityListElement0"
  >
    <div
      class="userAndMessage"
    >
      <span
        class="userWithIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon-regular"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
          />
        </svg>
        <p
          class="user"
        >
          user1@domain.com
        </p>
      </span>
      <p
        class="message"
      >
        Has logged in
      </p>
    </div>
    <div
      class="date"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root icon-regular"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
        <path
          d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
        />
      </svg>
      <p>
        Nov 28, 2019, 4:28 PM
      </p>
    </div>
  </div>
  <div
    class="row"
    data-testid="userActivityListElement1"
  >
    <div
      class="userAndMessage"
    >
      <span
        class="userWithIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon-regular"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
          />
        </svg>
        <p
          class="user"
        >
          user2@domain.com
        </p>
      </span>
      <p
        class="message"
      >
        Has logged in
      </p>
    </div>
    <div
      class="date"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root icon-regular"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
        <path
          d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
        />
      </svg>
      <p>
        Nov 27, 2019, 4:28 PM
      </p>
    </div>
  </div>
  <div
    class="row"
    data-testid="userActivityListElement2"
  >
    <div
      class="userAndMessage"
    >
      <span
        class="userWithIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon-regular"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
          />
        </svg>
        <p
          class="user"
        >
          user3@domain.com
        </p>
      </span>
      <p
        class="message"
      >
        Has logged in
      </p>
    </div>
    <div
      class="date"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root icon-regular"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
        <path
          d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
        />
      </svg>
      <p>
        Nov 26, 2019, 4:28 PM
      </p>
    </div>
  </div>
</div>
`;
